NullPointerException after upgrading to 0.29.0

<!-- Please fill out the following information to help us in analyzing the issue, but feel free to skip it if you don't have the information at hand or if it does not apply. Please remove everything which does not apply to the issue. -->

### Description
After upgrading from 0.28.0 to 0.29.0 we get the folloing NullPointerException when trying to build a docker image:
```
Caused by: java.lang.NullPointerException
    at io.fabric8.maven.docker.config.BuildImageConfiguration.getAbsoluteContextDirPath (BuildImageConfiguration.java:329)
    at io.fabric8.maven.docker.assembly.DockerAssemblyManager$1.customize (DockerAssemblyManager.java:133)
    at io.fabric8.maven.docker.assembly.DockerAssemblyManager.createBuildTarBall (DockerAssemblyManager.java:312)
    at io.fabric8.maven.docker.assembly.DockerAssemblyManager.createDockerTarArchive (DockerAssemblyManager.java:179)
    at io.fabric8.maven.docker.service.ArchiveService.createArchive (ArchiveService.java:123)
    at io.fabric8.maven.docker.service.ArchiveService.createArchive (ArchiveService.java:118)
    at io.fabric8.maven.docker.service.BuildService.buildImage (BuildService.java:117)
    at io.fabric8.maven.docker.service.BuildService.buildImage (BuildService.java:65)
    at io.fabric8.maven.docker.BuildMojo.buildAndTag (BuildMojo.java:73)
    at io.fabric8.maven.docker.BuildMojo.processImageConfig (BuildMojo.java:104)
    at io.fabric8.maven.docker.BuildMojo.executeInternal (BuildMojo.java:60)
    at io.fabric8.maven.docker.AbstractDockerMojo.execute (AbstractDockerMojo.java:225)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:954)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
```

Other than the version, we did not change any configuration.

### Info

* d-m-p version : 0.29.0
* Maven version (`mvn -v`) :
```
Apache Maven 3.5.4
Maven home: /usr/share/maven
Java version: 11.0.1, vendor: Oracle Corporation, runtime: /usr/lib/jvm/java-11-openjdk-amd64
Default locale: en_GB, platform encoding: UTF-8
OS name: "linux", version: "4.18.0-17-generic", arch: "amd64", family: "unix"
```
* Docker version : Docker version 18.06.1-ce, build e68fc7a

@papegaaij : Okay, so I tested it with your fix and NullPointerException was resolved. But there was some other error:

```
[INFO] DOCKER> Status: Image is up to date for docker.io/postgres:11
[INFO] DOCKER> Pulled postgres:11 in 6 seconds 
[INFO] Building tar: /home/Rohaan/work/repos/maven-docker-npe/target/docker/docker.topicusonderwijs.nl/keyhub/keyhub-db/14.1-SNAPSHOT/tmp/docker-build.tar
[INFO] DOCKER> [docker.topicusonderwijs.nl/keyhub/keyhub-db:14.1-SNAPSHOT]: Created docker-build.tar in 34 milliseconds
[INFO] DOCKER> Step 1/3 : FROM postgres:11
[INFO] DOCKER> ---> 9c116111eb08
[INFO] DOCKER> Step 2/3 : VOLUME /var/log/postgresql
[INFO] DOCKER> ---> Using cache
[INFO] DOCKER> ---> 4ee55faea159
[INFO] DOCKER> Step 3/3 : COPY createschema.sql /docker-entrypoint-initdb.d/
[ERROR] DOCKER> Unable to build image [docker.topicusonderwijs.nl/keyhub/keyhub-db:14.1-SNAPSHOT] : "lstat createschema.sql: no such file or directory"  ["lstat createschema.sql: no such file or directory" ]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 8.473 s
[INFO] Finished at: 2019-04-12T13:30:36+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.fabric8:docker-maven-plugin:0.29-SNAPSHOT:build (default-build) on project keyhub-docker-db: Unable to build image [docker.topicusonderwijs.nl/keyhub/keyhub-db:14.1-SNAPSHOT] : "lstat createschema.sql: no such file or directory"  -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

--- a/doc/changelog.md
+++ b/doc/changelog.md
@@ -3,6 +3,7 @@
 * **0.29-SNAPSHOT**
   - Restore ANSI color to Maven logging if disabled during plugin execution and enable color for Windows with Maven 3.5.0 or later. Color logging is enabled by default, but disabled if the Maven CLI disables color (e.g. in batch mode) ([#1108](https://github.com/fabric8io/docker-maven-plugin/issues/1108))
   - Fix NPE if docker:save is called with -Dfile=file-name-only.tar ([#1203](https://github.com/fabric8io/docker-maven-plugin/issues/1203))
+  - Fix NPE in BuildImageConfiguration ([#1200](https://github.com/fabric8io/docker-maven-plugin/issues/1200))
   - Improve GZIP compression performance for docker:save ([#1205](https://github.com/fabric8io/docker-maven-plugin/issues/1205))
 
 * **0.29.0** (2019-04-08)

--- a/pom.xml
+++ b/pom.xml
@@ -384,7 +384,7 @@
         <plugin>
           <groupId>org.apache.maven.plugins</groupId>
           <artifactId>maven-surefire-plugin</artifactId>
-	  <version>${surefire.version}</version>
+          <version>${surefire.version}</version>
           <configuration>
             <argLine>
               -javaagent:${settings.localRepository}/org/jmockit/jmockit/${jmockit.version}/jmockit-${jmockit.version}.jar

--- a/src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
+++ b/src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
@@ -162,7 +162,7 @@ public class BuildImageConfiguration implements Serializable {
     }
 
     public File getContextDir() {
-        return contextDir != null ? new File(contextDir) : getDockerFile().getParentFile();
+        return contextDir != null ? new File(contextDir) : getDockerFile().getAbsoluteFile().getParentFile();
     }
 
     public String getContextDirRaw() {
@@ -330,7 +330,16 @@ public class BuildImageConfiguration implements Serializable {
     }
 
     public File getAbsoluteDockerFilePath(MojoParameters mojoParams) {
-        return EnvUtil.prepareAbsoluteSourceDirPath(mojoParams, getDockerFile().getPath());
+        File absoluteDockerFilePath = EnvUtil.prepareAbsoluteSourceDirPath(mojoParams, getDockerFile().getPath());
+
+        // Update contextDir is dockerfile in src path is different than build config's.
+        if(!absoluteDockerFilePath.getAbsolutePath().equals(getDockerFile().getAbsolutePath())) {
+            dockerFile = absoluteDockerFilePath.getAbsolutePath();
+            if(contextDir == null) {
+                contextDir = absoluteDockerFilePath.getAbsoluteFile().getParent();
+            }
+        }
+        return absoluteDockerFilePath;
     }
 
     public File getAbsoluteDockerTarPath(MojoParameters mojoParams) {

--- a/src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
+++ b/src/main/java/io/fabric8/maven/docker/config/BuildImageConfiguration.java
@@ -174,7 +174,16 @@ public class BuildImageConfiguration implements Serializable {
     }
 
     public File getContextDir() {
-        return contextDir != null ? new File(contextDir) : getDockerFile().getAbsoluteFile().getParentFile();
+        if (!isDockerFileMode()) {
+            return null;
+        }
+        if (contextDir != null) {
+            return new File(contextDir);
+        }
+        if (getDockerFile().getParentFile() == null) {
+            return new File("");
+        }
+        return getDockerFile().getParentFile();
     }
 
     public String getContextDirRaw() {
@@ -342,16 +351,7 @@ public class BuildImageConfiguration implements Serializable {
     }
 
     public File getAbsoluteDockerFilePath(MojoParameters mojoParams) {
-        File absoluteDockerFilePath = EnvUtil.prepareAbsoluteSourceDirPath(mojoParams, getDockerFile().getPath());
-
-        // Update contextDir is dockerfile in src path is different than build config's.
-        if(!absoluteDockerFilePath.getAbsolutePath().equals(getDockerFile().getAbsolutePath())) {
-            dockerFile = absoluteDockerFilePath.getAbsolutePath();
-            if(contextDir == null) {
-                contextDir = absoluteDockerFilePath.getAbsoluteFile().getParent();
-            }
-        }
-        return absoluteDockerFilePath;
+        return EnvUtil.prepareAbsoluteSourceDirPath(mojoParams, getDockerFile().getPath());
     }
 
     public File getAbsoluteDockerTarPath(MojoParameters mojoParams) {

--- a/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
@@ -60,6 +60,18 @@ public class BuildImageConfigurationTest {
     }
 
     @Test
+    // Tests fix for #1200
+    public void simpleDockerfileWithoutParentDir() {
+        BuildImageConfiguration config =
+            new BuildImageConfiguration.Builder().
+                dockerFile("Dockerfile").build();
+        config.initAndValidate(logger);
+        assertTrue(config.isDockerFileMode());
+        assertEquals(config.getDockerFile(),new File("Dockerfile"));
+        assertTrue(config.getContextDir().isAbsolute());
+    }
+
+    @Test
     public void simpleDockerfileDir() {
         BuildImageConfiguration config =
             new BuildImageConfiguration.Builder().

--- a/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
@@ -56,7 +56,7 @@ public class BuildImageConfigurationTest {
         config.initAndValidate(logger);
         assertTrue(config.isDockerFileMode());
         assertEquals(config.getDockerFile(),new File("src/docker/Dockerfile"));
-        assertEquals(config.getContextDir(),new File("src/docker"));
+        assertTrue(config.getContextDir().isAbsolute());
     }
 
     @Test
@@ -67,7 +67,7 @@ public class BuildImageConfigurationTest {
         config.initAndValidate(logger);
         assertTrue(config.isDockerFileMode());
         assertEquals(config.getDockerFile(),new File("src/docker/Dockerfile"));
-        assertEquals(config.getContextDir(),new File("src/docker"));
+        assertTrue(config.getContextDir().isAbsolute());
     }
 
     @Test

--- a/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationTest.java
@@ -52,11 +52,11 @@ public class BuildImageConfigurationTest {
     public void simpleDockerfile() {
         BuildImageConfiguration config =
             new BuildImageConfiguration.Builder().
-                dockerFile("src/docker/Dockerfile").build();
+                dockerFile("src/main/docker/Dockerfile").build();
         config.initAndValidate(logger);
         assertTrue(config.isDockerFileMode());
-        assertEquals(config.getDockerFile(),new File("src/docker/Dockerfile"));
-        assertTrue(config.getContextDir().isAbsolute());
+        assertEquals(config.getDockerFile(),new File("src/main/docker/Dockerfile"));
+        assertEquals(config.getContextDir(),new File("src/main/docker"));
     }
 
     @Test
@@ -68,8 +68,9 @@ public class BuildImageConfigurationTest {
         config.initAndValidate(logger);
         assertTrue(config.isDockerFileMode());
         assertEquals(config.getDockerFile(),new File("Dockerfile"));
-        assertTrue(config.getContextDir().isAbsolute());
+        assertEquals(config.getContextDir(), new File(""));
     }
+
 
     @Test
     public void simpleDockerfileDir() {
@@ -79,7 +80,7 @@ public class BuildImageConfigurationTest {
         config.initAndValidate(logger);
         assertTrue(config.isDockerFileMode());
         assertEquals(config.getDockerFile(),new File("src/docker/Dockerfile"));
-        assertTrue(config.getContextDir().isAbsolute());
+        assertFalse(config.getContextDir().isAbsolute());
     }
 
     @Test

--- a/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationWithMavenDepsTest.java
+++ b/src/test/java/io/fabric8/maven/docker/config/BuildImageConfigurationWithMavenDepsTest.java
@@ -1,0 +1,60 @@
+package io.fabric8.maven.docker.config;
+
+import java.io.File;
+
+import io.fabric8.maven.docker.util.Logger;
+import io.fabric8.maven.docker.util.MojoParameters;
+import mockit.Expectations;
+import mockit.Mocked;
+import org.apache.maven.project.MavenProject;
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author roland
+ * @since 2019-04-21
+ */
+public class BuildImageConfigurationWithMavenDepsTest {
+
+    @Mocked
+    private MojoParameters params;
+
+    @Mocked
+    private MavenProject project;
+
+    @Mocked
+    Logger logger;
+
+    @Test
+    public void simpleDockerfileWithoutParentDir() {
+        new Expectations() {{
+           params.getSourceDirectory(); result = "src/main/docker";
+           params.getProject(); result = project;
+           project.getBasedir(); result = "/project";
+        }};
+
+        String[] data = new String[] {
+            null, "Dockerfile", "/project/src/main/docker", "/project/src/main/docker/Dockerfile",
+            null, "/context/Dockerfile", "/context", "/context/Dockerfile",
+            "/context", "Dockerfile", "/context", "/context/Dockerfile",
+            "context", "Dockerfile", "/project/src/main/docker/context", "/project/src/main/docker/context/Dockerfile",
+            "context", "/other/Dockerfile", "/project/src/main/docker/context", "/other/Dockerfile",
+            "/context", "/other/Dockerfile", "/context", "/other/Dockerfile"
+        };
+
+        for (int i = 0; i < data.length; i+= 4) {
+            BuildImageConfiguration config =
+                new BuildImageConfiguration.Builder()
+                    .contextDir(data[i])
+                    .dockerFile(data[i + 1]).build();
+            config.initAndValidate(logger);
+
+
+            assertEquals(data[i + 2], config.getAbsoluteContextDirPath(params).getAbsolutePath());
+            assertEquals(data[i + 3], config.getAbsoluteDockerFilePath(params).getAbsolutePath());
+        }
+    }
+
+}

